### ISSUE TITLE
Bug with adding self-edges to graphs?

### ISSUE THREAD

**Issue by mkhorton** [pygraphviz/pygraphviz#133]
To illustrate this bug, this snippet works as expected:

    A=pygraphviz.AGraph(strict=False, directed=True)
    A.add_node(0)
    A.add_node(1)
    A.add_edge(0, 0, key="1")
    A.add_edge(0, 1, key="2")
    A.add_edge(0, 1, key="1")
    A.add_edge(0, 0, key="2")
    print(A.edges(keys=True))  # prints [('0', '0', '1'), ('0', '0', '2'), ('0', '1', '2'), ('0', '1', '1')]

While this snippet does not:

    A=pygraphviz.AGraph(strict=False, directed=True)
    A.add_node(0)
    A.add_node(1)
    A.add_edge(0, 0, key="1")
    A.add_edge(0, 0, key="2")
    A.edges(keys=True) # prints [('0', '0', '1')] and *not* [('0', '0', '1'), ('0', '0', '2')] as expected

In other words, adding multiple self-edges (here 0 to 0) seems to work inconsistently.

What's going on here? Many thanks for any help!

(For context, I discovered this when trying to use [networkx's `to_agraph`](https://networkx.github.io/documentation/networkx-1.9/_modules/networkx/drawing/nx_agraph.html#to_agraph) function which uses pygraphviz and noticing inconsistent results.)

**Comment by mkhorton** [comment]
To provide another example:

    A=pygraphviz.AGraph(strict=False, directed=True)
    A.add_node(0)
    A.add_node(1)
    A.add_edge(0, 0, key="1")
    A.add_edge(0, 1, key="2")
    A.add_edge(0, 0, key="2")
    A.remove_edge(0, 1, key="2")
    A.edges(keys=True)  # returns [('0', '0', '1'), ('0', '0', '2')]

But the exact same example, swapping the last two lines, does not:

    A=pygraphviz.AGraph(strict=False, directed=True)
    A.add_node(0)
    A.add_node(1)
    A.add_edge(0, 0, key="1")
    A.add_edge(0, 1, key="2")
    A.remove_edge(0, 1, key="2")
    A.add_edge(0, 0, key="2")
    A.edges(keys=True)  # returns [('0', '0', '1')]

**Comment by hagberg** [comment]
This may be an issue with your graphviz version or compiled library.  It does work correctly, e.g.
```
In [1]: import pygraphviz

In [2]: A=pygraphviz.AGraph(strict=False, directed=True)
   ...: A.add_node(0)
   ...: A.add_node(1)
   ...: A.add_edge(0, 0, key="1")
   ...: A.add_edge(0, 1, key="2")
   ...: A.add_edge(0, 0, key="2")
   ...: A.remove_edge(0, 1, key="2")
   ...: A.edges(keys=True)  # returns [('0', '0', '1'), ('0', '0', '2')]
   ...: 
Out[2]: [('0', '0', '1'), ('0', '0', '2')]

In [3]: A=pygraphviz.AGraph(strict=False, directed=True)
   ...: A.add_node(0)
   ...: A.add_node(1)
   ...: A.add_edge(0, 0, key="1")
   ...: A.add_edge(0, 1, key="2")
   ...: A.remove_edge(0, 1, key="2")
   ...: A.add_edge(0, 0, key="2")
   ...: A.edges(keys=True)  # returns [('0', '0', '1')]
   ...: 
Out[3]: [('0', '0', '1'), ('0', '0', '2')]

In [4]: pygraphviz.__version__
Out[4]: '1.5.dev'
```

My configuration is (Linux)
```
$ python --version
Python 3.6.1 :: Anaconda 4.4.0 (64-bit)
$ dot -V
dot - graphviz version 2.38.0 (20140413.2041)
```
It could be triggering this bug https://github.com/pygraphviz/pygraphviz/issues/111 which has been fixed upstream in the graphviz library https://github.com/ellson/graphviz/issues/1252

**Comment by mkhorton** [comment]
Interesting, thank you for your swift reply!

My own configuration is:

    $ python --version
    Python 3.6.1
    $ dot -V
    dot - graphviz version 2.40.1 (20161225.0304)
    $ python -c 'import pygraphviz; print(pygraphviz.__version__)' 
    1.3.1

All installed via Homebrew on macOS 10.12.6, graphviz 2.40.1 is the latest on there. Weirdly, I'm using a later version of graphviz than you are, so maybe it's a regression?

**Comment by hagberg** [comment]
Yes, the graphviz bug is in 2.40.1 and not 2.38.0.

**Comment by mkhorton** [comment]
Thanks -- I can confirm this too. For any homebrew users, installing 2.38.0 via [this commit](https://github.com/Homebrew/homebrew-core/blob/7a4dabfc1a2acd9f01a1670fde4f0094c4fb6ffa/Formula/graphviz.rb) seems to fix this.
